source/distributed/fully_distributed_tria.cc

```diff
@@ -301,8 +313,140 @@ namespace parallel
     void
     Triangulation<dim, spacedim>::execute_coarsening_and_refinement()
     {
-      Assert(false, ExcNotImplemented());
-    }
+      // 1) preparation
+      {
+        // reset the coarsening/refinement flags of non-local cells
+        for (auto cell : this->cell_iterators())
+          if (cell->active() && !cell->is_locally_owned())
+            {
+              cell->clear_refine_flag();
+              cell->clear_coarsen_flag();
+
+              // start assumption:
+              if (cell->level() > 0)
+                cell->set_coarsen_flag();
+            }
+
+        // update refinement/coarsening flags of non-local cells via
+        // ghost-cell exchange
+        GridTools::exchange_cell_data_to_ghosts<unsigned int>(
+          *this,
+          [](const TriaIterator<CellAccessor<dim, spacedim>> &cell) {
+            if (cell->refine_flag_set())
+              {
+                Assert(cell->refine_flag_set() ==
+                         RefinementPossibilities<dim>::isotropic_refinement,
+                       ExcMessage(
+                         "This class does not support anisotropic refinement"));
+                return true;
+              }
+            else
+              return false;
+          },
+          [](const TriaIterator<CellAccessor<dim, spacedim>> &cell,
+             const unsigned int                               refinement_flag) {
+            if (refinement_flag == true)
+              {
+                cell->clear_coarsen_flag();
```

Comment:
how do you get by with just a single bool, shouldn't you have refine|coarsen|keep ?